---
title: "CT-Pop Biomarker Density Graphs Notebook"
output:
  # html_document:
  # df_print: paged
  pdf_document: 
    dev: pdf
    fig_caption: no
    number_sections: yes
    toc: yes
    toc_depth: 3
documentclass: article
urlcolor: blue
classoption: a4paper
header-includes:
- \usepackage{graphicx}
- \usepackage{float}
---

# Environment 

```{r environment, message=FALSE, warning=FALSE}
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)
library(magrittr)
library(ggplot2)
library(ggridges)

options(scipen = 999)

knitr::opts_chunk$set(
    echo = FALSE,
    message = FALSE,
    warning = FALSE,
    fig.align='center',
    fig.pos='H',
    fig.path = "../output/densityplots/",
    dev = c("pdf"),
    dpi=500
)
```

# Load Data

```{r load}
df <- read.csv("https://raw.githubusercontent.com/x-atlas-consortia/hra-pop/main/output-data/v0.10.2/reports/atlas/validation-b-mean-expression.csv",
               header=T, fileEncoding = "UTF8")
```

# Prepare data

```{r prep}
# Removes proteomics tools
df <- df[df$tool != "sc_proteomics",]

# Names alignment
names(df)[4] <- "organ"
names(df)[7] <- "cell"
names(df)[8] <- "gene"

# Prepare organ labels
df$organ <- df$organ %>% 
  str_replace(" of body", "") %>% 
  str_to_title()

# Factor organs
df$organ <- factor(df$organ)

# Prepare Cell Labels
df$cell <- 
  df$cell %>%
  str_to_upper()

# Prepare tool
df$tool <- df$tool %>%
  str_to_title() 

# Factor tools
df$tool <- factor(df$tool)
```

# Rank biomarkers and subsetting data 

```{r ranking}
# Count the number cells with a biomarker measurements for each organ
df <- df %>% 
  arrange(organ, gene) %>%
  select(organ, gene, cell, mean_gene_expression_value) %>%
  group_by(organ, gene) %>%
  mutate(ct=row_number())

# Ranks top 20 biomarkers based on the number measurements for each organ
gene_ranks_o <- df %>%
  ddply(.(organ, gene), summarise,
        max_cells = max(ct)) %>%
  arrange(organ, desc(max_cells)) %>%
  group_by(organ) %>%
  mutate(rank=row_number()) %>%
  filter(rank < 21)
 
# Selects all relevant measurements for top 20 biomarker measurements by organ
gene_sets <- right_join(df, gene_ranks_o,by=c("organ","gene")) %>%
  arrange(organ, rank)

# Identify top 40 biomarkers across all data 
gene_ranks_f <- gene_sets %>%
  arrange(desc(max_cells), gene)
gene_ranks_f <- unique(gene_ranks_f[,c(1,2,6)])
gene_ranks_f$rank_o <- as.numeric(rownames(gene_ranks_f))
gene_ranks_f <- gene_ranks_f[gene_ranks_f$rank_o < 41,]

# Selects all relevant measurements for top 40 biomarkers across all data.
gene_sets_o <- right_join(gene_sets, gene_ranks_f[,c(1,2,4)],by=c("organ","gene")) %>%
  arrange(rank_o)
gene_sets_o$organ <- factor(as.character(gene_sets_o$organ))
gene_sets_o$gene <- factor(gene_sets_o$gene)
```

# Visualization

```{r densityplot_biomarkers_organ, fig.height=10, fig.width=6, message=FALSE}
theme_set(theme_minimal())

ggplot(gene_sets_o, aes(x=log(mean_gene_expression_value), 
                        y=gene)) +
  geom_density_ridges(aes(fill=organ), 
                      rel_min_height = 0.01, alpha=.5) +
  #facet_wrap(vars(organ)) +
  labs(x="Mean Gene Expression Values (log2)", y="Biomarker") +
  scale_fill_discrete("Organ") +
  theme(legend.justification = "top")

```
























