---
title: "CT-Pop Biomarker Density Graphs Notebook"
output: html_document
---

# Environment 

```{r environment, message=FALSE, warning=FALSE}
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)
library(magrittr)
library(ggplot2)
library(ggridges)

options(scipen = 999)
```

# Load Data

```{r load}
df <- read.csv("https://raw.githubusercontent.com/x-atlas-consortia/hra-pop/main/output-data/v0.10.2/reports/atlas/validation-b-mean-expression.csv",
               header=T, fileEncoding = "UTF8")
```

# Prepare data

```{r prep}
# Removes proteomics tools
df <- df[df$tool != "sc_proteomics",]

# Names alignment
names(df)[4] <- "organ"
names(df)[7] <- "cell"
names(df)[8] <- "gene"

# Prepare organ labels
df$organ <- df$organ %>% 
  str_replace(" of body", "") %>% 
  str_to_title()

# Factor organs
df$organ <- factor(df$organ)

# Prepare Cell Labels
df$cell <- 
  df$cell %>%
  str_to_upper()

# Prepare tool
df$tool <- df$tool %>%
  str_to_title() 

# Factor tools
df$tool <- factor(df$tool)
```

# Rank biomarkers and subsetting data 

```{r ranking}
# Count the number cells with a biomarker measurements for each organ
df <- df %>% 
  arrange(organ, gene) %>%
  select(organ, gene, cell, mean_gene_expression_value) %>%
  group_by(organ, gene) %>%
  mutate(ct=row_number())

# Ranks top 20 biomarkers based on the number measurements for each organ
gene_ranks_o <- df %>%
  ddply(.(organ, gene), summarise,
        max_cells = max(ct)) %>%
  arrange(organ, desc(max_cells)) %>%
  group_by(organ) %>%
  mutate(rank=row_number()) %>%
  filter(rank < 21)
 
# Selects all relevant measurements for top 20 biomarker measurements by organ
gene_sets <- right_join(df, gene_ranks_o,by=c("organ","gene")) %>%
  arrange(organ, rank)

# Identify top 40 biomarkers across all data 
gene_ranks_f <- 
  gene_sets %>%
  arrange(organ, desc(max_cells), gene)
gene_ranks_f <- 
  unique(gene_ranks_f[,c(1,2,6)]) %>%
  group_by(organ) %>%
  mutate(rank_organ = row_number()) %>%
  ungroup() %>%
  arrange(desc(max_cells)) %>%
  mutate(rank_overall = row_number())

# Selects all relevant measurements for top 40 biomarkers across all data.
gene_ranks_f1 <- gene_ranks_f[gene_ranks_f$rank_overall < 41,]

## Joins gene sets data with overall ranks 
gene_sets_o1 <- 
  right_join(gene_sets, gene_ranks_f1[,c(1,2,4,5)],
                          by=c("organ","gene")) %>%
  arrange(rank_overall)

## Factoring variables
gene_sets_o1$organ <- 
  factor(as.character(gene_sets_o1$organ),
         levels=c("Heart","Lung","Kidney","Small Intestine","Large Intestine"))
gene_sets_o1$gene <- 
  factor(gene_sets_o1$gene)

# Selects all data of top 20 biomarkers for each organ
gene_ranks_f2 <- gene_ranks_f[gene_ranks_f$rank_organ < 21,]

## Joins gene sets data with organ gene ranks 
gene_sets_o2 <- 
  right_join(gene_sets, gene_ranks_f2[,c(1,2,4,5)],
                          by=c("organ","gene")) %>%
  arrange(rank_overall)

## Factoring variables
gene_sets_o2$organ <- 
  factor(as.character(gene_sets_o2$organ),
         levels=c("Heart","Lung","Kidney","Small Intestine","Large Intestine"))
```

### Visualization

#### Overall - Top 20 Biomarkers (cell counts) - Gene Expression Density Ridge Plots

```{r densityplot_biomarkers_organ_overall, fig.height=10, fig.width=6, message=FALSE, warning=FALSE}
theme_set(theme_minimal())

color_pal_kid2 <- c("#E97B74","#F9CE8D","#A295C9","#7495AF","#8DC599")
                  
overall_dr <- ggplot(gene_sets_o1, 
                     aes(x=log(mean_gene_expression_value), y=gene)) +
                  geom_density_ridges(aes(fill=organ),
                                      bandwidth = 0.15,
                                      rel_min_height = 0.01, alpha=.55) +
                  labs(x="Mean Gene Expression Values (log2)", y="Biomarker") +
                  scale_fill_manual(name="Organ", values=color_pal_kid2) +
                  theme(legend.justification = "top")

# Export visualization
ggsave(file=paste0("../output/densityplots/overall_top20_gene_expression_v0.1.pdf"),
       overall_dr, width=5, height=6.5, units="in")
overall_dr
```


#### Heart - Top 20 Biomarkers (cell counts) - Gene Expression Density Ridge Plots
```{r density_heart, fig.height=6, fig.width=4.75, message=FALSE, warning=FALSE}
theme_set(theme_minimal())

# Subset organ
heart <- gene_sets_o2[gene_sets_o2$organ=="Heart",] 

# Unique factor with cell count
labs <- unique(heart[order(heart$max_cells),c(2,6)]) %>%
  mutate(facs = paste0(gene," (",max_cells,")")) %>%
  filter(gene != "NA") 
heart <- left_join(heart,labs[,c(1,3)],by="gene")
heart$facs <- factor(heart$facs, 
                     levels=labs$facs)

# Color palette 
color_org <- c("#E97B74")

# Visualization
hrt_dr <- heart %>% 
            filter(facs != "NA") %>%
            ggplot(aes(x=log(mean_gene_expression_value), 
                       y=facs)) +
              geom_density_ridges(aes(fill=organ), 
                                  na.rm = TRUE,,
                                  bandwidth = 0.15,
                                  rel_min_height = 0.01, alpha=.85) +
              labs(x="Mean Gene Expression Values (log2)", y="Biomarker (Cells)") +
              scale_fill_manual(name="Organ", values=color_org) +
              theme(legend.position = "none")

# Export pdf
ggsave(file=paste0("../output/densityplots/heart_top20_gene_expression_v0.1.pdf"),
       hrt_dr, width=4.75, height=6, units="in")
hrt_dr
```


#### Lung - Top 20 Biomarkers (cell counts) - Gene Expression Density Ridge Plots
```{r density_lung, fig.height=6, fig.width=4.75, message=FALSE, warning=FALSE}
theme_set(theme_minimal())

# Subset organ
lung <- gene_sets_o2[gene_sets_o2$organ=="Lung",] 

# Unique factor with cell count
labs <- unique(lung[order(lung$max_cells),c(2,6)]) %>%
  mutate(facs = paste0(gene," (",max_cells,")")) %>%
  filter(gene != "NA") 
lung <- left_join(lung,labs[,c(1,3)],by="gene")
lung$facs <- factor(lung$facs, 
                    levels=labs$facs)

# Color palette 
color_org <- c("#F9CE8D")

# Visualization
lung_dr <- lung %>% 
            filter(facs != "NA") %>%
            ggplot(aes(x=log(mean_gene_expression_value), 
                       y=facs)) +
              geom_density_ridges(aes(fill=organ), 
                                  na.rm = TRUE,,
                                  bandwidth = 0.15,
                                  rel_min_height = 0.01, alpha=.85) +
              labs(x="Mean Gene Expression Values (log2)", y="Biomarker") +
              scale_fill_manual(name="Organ", values=color_org) +
              theme(legend.position = "none")

# Export pdf
ggsave(file=paste0("../output/densityplots/lung_top20_gene_expression_v0.1.pdf"),
       lung_dr, width=4.75, height=6, units="in")
lung_dr
```


#### Kidney - Top 20 Biomarkers (cell counts) - Gene Expression Density Ridge Plots
```{r density_kidney, fig.height=6, fig.width=4.75, message=FALSE, warning=FALSE}
# Subset organ
kid <- gene_sets_o2[gene_sets_o2$organ=="Kidney",] 

# Unique factor with cell count
labs <- unique(kid[order(kid$max_cells),c(2,6)]) %>%
  mutate(facs = paste0(gene," (",max_cells,")")) %>%
  filter(gene != "NA") 
kid <- left_join(kid,labs[,c(1,3)],by="gene")
kid$facs <- factor(kid$facs, 
                   levels=labs$facs)

# Color palette 
color_org <- c("#A295C9")
#"#7495AF","#8DC599"

# Visualization
kid_dr <- kid %>% 
            filter(facs != "NA") %>%
            ggplot(aes(x=log(mean_gene_expression_value), 
                       y=facs)) +
              geom_density_ridges(aes(fill=organ), 
                                  na.rm = TRUE,,
                                  bandwidth = 0.15,
                                  rel_min_height = 0.01, alpha=.85) +
              labs(x="Mean Gene Expression Values (log2)", y="Biomarker") +
              scale_fill_manual(name="Organ", values=color_org) +
              theme(legend.position = "none")

# Export pdf
ggsave(file=paste0("../output/densityplots/kidney_top20_gene_expression_v0.1.pdf"),
       kid_dr, width=4.75, height=6, units="in")
kid_dr
```

#### Small Intestine - Top 20 Biomarkers (cell counts) - Gene Expression Density Ridge Plots
```{r density_sml, fig.height=6, fig.width=4.75, message=FALSE, warning=FALSE}
# Subset organ
sml <- gene_sets_o2[gene_sets_o2$organ=="Small Intestine",] 

# Unique factor with cell count
labs <- unique(sml[order(sml$max_cells),c(2,6)]) %>%
  mutate(facs = paste0(gene," (",max_cells,")")) %>%
  filter(gene != "NA") 
sml <- left_join(sml,labs[,c(1,3)],by="gene")
sml$facs <- factor(sml$facs, 
                   levels=labs$facs)

# Color palette 
color_org <- c("#7495AF")

# Visualization
sml_dr <- sml %>% 
            filter(facs != "NA") %>%
            ggplot(aes(x=log(mean_gene_expression_value), 
                       y=facs)) +
              geom_density_ridges(aes(fill=organ), 
                                  na.rm = TRUE,,
                                  bandwidth = 0.15,
                                  rel_min_height = 0.01, alpha=.85) +
              labs(x="Mean Gene Expression Values (log2)", y="Biomarker") +
              scale_fill_manual(name="Organ", values=color_org) +
              theme(legend.position = "none")

# Export pdf
ggsave(file=paste0("../output/densityplots/smallIntestine_top20_gene_expression_v0.1.pdf"),
       sml_dr, width=4.75, height=6, units="in")
sml_dr
```

#### Large Intestine - Top 20 Biomarkers (cell counts) - Gene Expression Density Ridge Plots
```{r density_lrg, fig.height=6, fig.width=4.75, message=FALSE, warning=FALSE}
# Subset organ
lrg <- gene_sets_o2[gene_sets_o2$organ=="Large Intestine",] 

# Unique factor with cell count
labs <- unique(lrg[order(lrg$max_cells),c(2,6)]) %>%
  mutate(facs = paste0(gene," (",max_cells,")")) %>%
  filter(gene != "NA") 
lrg <- left_join(lrg,labs[,c(1,3)],by="gene")
lrg$facs <- factor(lrg$facs, 
                   levels=labs$facs)

# Color palette 
color_org <- c("#8DC599")

# Visualization
lrg_dr  <- lrg %>% 
            filter(facs != "NA") %>%
            ggplot(aes(x=log(mean_gene_expression_value), 
                       y=facs)) +
              geom_density_ridges(aes(fill=organ), 
                                  na.rm = TRUE,,
                                  bandwidth = 0.15,
                                  rel_min_height = 0.01, alpha=.85) +
              labs(x="Mean Gene Expression Values (log2)", y="Biomarker") +
              scale_fill_manual(name="Organ", values=color_org) +
              theme(legend.position = "none")

# Export pdf
ggsave(file=paste0("../output/densityplots/largeIntestine_top20_gene_expression_v0.1.pdf"),
       lrg_dr, width=4.75, height=6, units="in")
lrg_dr
```












